[Store] 매장 전체 조회 기능 구현

iop0217 opened the pull request:
### Description  
- 매장 전체 목록 조회 기능을 구현했습니다.  
- 가게 이름 검색 기능(부분 일치)과 페이지네이션(page, size) 기능을 적용했습니다.  
- 매장이 없는 경우에도 404가 아닌 빈 배열과 200 OK를 반환하도록 처리했습니다.  
- Postman 테스트를 통해 정상 동작을 확인했습니다.
- **PR 리뷰 반영하여 수정 완료했습니다.**

---

### Changes  
- `GET /api/stores` 매장 전체 조회 API 구현
- 검색어(가게 이름) 부분 검색 기능 적용
- 페이지네이션 기능 적용 (`page`, `size` 파라미터)
- 매장이 없는 경우 200 OK와 빈 배열 반환
- Postman을 통한 정상 동작 검증 완료
- **@PageableDefault(page=0, size=10) 적용 및 구조 간결화**
- **page/size 입력값 보정**
- **매직 넘버 제거 및 상수화**

---

### Screenshots  

#### ✅ 200 OK
![200](https://github.com/user-attachments/assets/f9178439-ba66-4d21-94ed-3b248a6927f6)

#### ❌ 401 UNAUTHORIZED
![401](https://github.com/user-attachments/assets/98dc4490-b6ab-46f0-8c69-6ac8530637f9)

iop0217 committed:
feature[store]: 매장 전체 조회 기능 및 테스트 코드 구현

- 검색어 유무에 따른 매장 조회 처리
- 전체 조회 테스트 코드 작성

iop0217 committed:
fix[store]: 매장 전체 조회 API 수정

iop0217 committed:
Merge branch 'dev' into feature/suhyun-stores

iop0217 committed:
fix[store]: 페이지네이션 유효성 검증 추가

- page가 0 미만일 경우 GlobalException 발생
- size가 1 미만 또는 100 초과일 경우 GlobalException 발생
- Postman 테스트를 통해 정상 동작 확인

iop0217 committed:
fix[store]: 에러 메시지 수정

iop0217 committed:
fix[store]: 페이지 검증 수정

iop0217 committed:
fix[store]: GlobalException 처리 메서드 추가

iop0217 committed:
fix[store]: 페이즈 및 사이즈 검증 Exception 수정

iop0217 committed:
fix[store]: 페이지 및 사이즈 검증 코드 수정

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/store/controller/StoreController.java:
```diff
@@ -30,4 +34,25 @@ public ResponseEntity<StoreResponseDto> createStore(
+            @RequestParam(name = "page", defaultValue = "0") int page,
+            @RequestParam(name = "size", defaultValue = "10") int size
+    ) {
+        Pageable pageable = PageRequest.of(page, size);
+
+        Page<StoreListResponseDto> storePage = storeService.getStores(storeName, pageable);
+
+        Map<String, Object> result = new LinkedHashMap<>();
+        result.put("stores", storePage.getContent());
+        result.put("totalElements", storePage.getTotalElements());
+        result.put("page", storePage.getNumber());
+        result.put("size", storePage.getSize());
+        result.put("totalPages", storePage.getTotalPages());
+
+        return ResponseEntity.ok(result);
```
Page객체를 그대로 응답하지 않고, 클라이언트에서 필요한 정보만 보내는것은 매우 괜찮은 선택이라고 생각합니다만,
이렇게 되면 모든 전체 조회 api마다 각기 다른 응답을 보낼 가능성이 높기 때문에 공통적으로 사용하는 Page 객체가 하나 있는게 더 좋을 것 같다는 생각이 듭니다.

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import java.util.List;

@Getter
@NoArgsConstructor
@AllArgsConstructor
public class PageResponse<T> {

    private List<T> content;
    private long totalElements;
    private int page;
    private int size;
    private int totalPages;

    public static <T> PageResponse<T> from(Page<T> page) {
        return new PageResponse<>(
                page.getContent(),
                page.getTotalElements(),
                page.getNumber(),
                page.getSize(),
                page.getTotalPages()
        );
    }
}

요런 클래스는 common 패키지에 두고 모두가 응답할때 해당 객체를 사용하면 통일된 응답을 줄 수 있을 것 같네요

iop0217 replied:
수정했습니다

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/store/controller/StoreController.java:
```diff
@@ -30,4 +34,25 @@ public ResponseEntity<StoreResponseDto> createStore(
+
+    // 매장 전체 조회
+    @GetMapping
+    public ResponseEntity<Map<String, Object>> getStores(
+            @RequestParam(name = "storeName", defaultValue = "") String storeName,
+            @RequestParam(name = "page", defaultValue = "0") int page,
+            @RequestParam(name = "size", defaultValue = "10") int size
+    ) {
+        Pageable pageable = PageRequest.of(page, size);
```
Pageable을 사용하지 않고 쿼리파라미터로 받으시길래 사용안하시는 줄 알았는데 내부에서 Pageable을 생성하시더라구요
그냥 컨트롤러에서 파라미터로 Pageable pageable을 받으면 클라이언트에서 page, size로 전달하면 Pageable로 자동 바인딩 됩니다.

 @PageableDefault(page = 0, size = 10)
 이런식으로 설정할 수 있는데 page=0, size=10이 기본값이라 그대로 용하실꺼면 애노테이션도 필요 없어지니 가독성도 좋아지구요

iop0217 replied:
원래 비슷한 방식으로 하다가 postman에서 의도한 오류가 발생하지 않아서 쿼리로 만들었어요 다시 해볼게요

iop0217 replied:
수정했습니다

nagul2 commented on src/main/java/com/delivery/igo/igo_delivery/api/store/service/StoreServiceImpl.java:
```diff
@@ -36,10 +39,39 @@ public StoreResponseDto createStore(StoreRequestDto requestDto, Long userId) {
+        int page = pageable.getPageNumber();
+        int size = pageable.getPageSize();
+
+        // 잘못된 페이지 번호인 경우
+        if (page < 0) {
+            throw new GlobalException(ErrorCode.INVALID_PAGE_PARAMETER);
+        }
+
+        // 잘못된 사이즈 크기인 경우
+        if (size <= 0) {
+            throw new GlobalException(ErrorCode.INVALID_SIZE_TOO_SMALL);
+        }
+
+        // 사이즈 크기가 너무 큰 경우
+        if (size > 100) {
+            throw new GlobalException(ErrorCode.INVALID_SIZE_TOO_LARGE);
+        }
+
```
이부분은 에러를 던지기보단 기본값으로 반환하는게 더 사용자 친화적이지 않을까 생각됩니다.

예를들어 page가 음수로 떨어지면 0으로 고정, size가 음수로 떨어지면 절대값으로 출력한다든가 기본값으로 고정,
사이즈가 너무 큰경우에는 에러가 필요한지는 잘 모르겠습니다.

iop0217 replied:
수정했습니다. 너무 큰 경우는 메모리 때문에 걸어뒀습니다. 실무에서는 100~500 정도로 제한한다고 하여 500으로 늘려놨습니다.

anyubinn commented on src/main/java/com/delivery/igo/igo_delivery/api/store/service/StoreServiceImpl.java:
```diff
@@ -36,10 +39,39 @@ public StoreResponseDto createStore(StoreRequestDto requestDto, Long userId) {
+        int size = pageable.getPageSize();
+
+        // 잘못된 페이지 번호인 경우
+        if (page < 0) {
```
페이지 번호와 사이즈 검증 부분에서 매직 넘버가 사용되고 있는 것 같아요! 의미를 명확히 하기 위해 상수로 분리하는 것도 고려해보면 좋을 것 같아요!

iop0217 replied:
수정했습니다

iop0217 committed:
Merge branch 'dev' into feature/suhyun-stores

iop0217 committed:
refactor[store]: 매장 전체 조회 페이지네이션 로직 PR 반영

- @PageableDefault(page=0, size=10) 적용하여 기본 페이지네이션 설정
- 잘못된 page/size 입력값에 대한 보정 로직 유지
- 매직 넘버 제거 및 의미 있는 상수(MIN_PAGE_SIZE, DEFAULT_PAGE_SIZE, MAX_PAGE_SIZE)로 관리
- 컨트롤러 구조 간소화 및 코드 가독성 개선
- PR 피드백을 반영하여 최종 정상 동작 확인

iop0217 committed:
refactor[store]: 페이지네이션 최대 사이즈 500 → 50 수정

iop0217 committed:
Merge branch 'dev' into feature/suhyun-stores
